rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\DowngradePhp70\Rector\Expr;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\ArrayDimFetch;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\PropertyFetch;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use Rector\Core\Rector\AbstractRector;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://wiki.php.net/rfc/uniform_variable_syntax
+ *
+ * @see \Rector\Tests\DowngradePhp70\Rector\Expr\DowngradeUnnecessarilyParenthesizedExpressionRector\DowngradeUnnecessarilyParenthesizedExpressionRectorTest
+ */
+final class DowngradeUnnecessarilyParenthesizedExpressionRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Remove parentheses around expressions where they are not necessary.',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+($f)['foo'];
+($f)->foo;
+($f)->foo();
+($f)::$foo;
+($f)::foo();
+($f)();
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+$f['foo'];
+$f->foo;
+$f->foo();
+$f::$foo;
+$f::foo();
+$f();
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        // TODO: Make PHPStan allow Expr namespace.
+        $dummyClassNameUse = str_contains(Expr::class, '\\');
+        assert($dummyClassNameUse);
+        return [
+            ArrayDimFetch::class,
+            PropertyFetch::class,
+            MethodCall::class,
+            StaticPropertyFetch::class,
+            StaticCall::class,
+            FuncCall::class,
+        ];
+    }
+
+    /**
+     * @param ArrayDimFetch|PropertyFetch|MethodCall|StaticPropertyFetch|StaticCall|FuncCall $node
+     */
+    public function refactor(Node $node): Expr
+    {
+        // Uniform variable syntax RFC allows wrapping arbitrary expressions in parentheses.
+        // Since PHP 5 cannot parse that we need to strip them.
+        // And while parenthesization is not part of the AST, Rector only re-generates
+        // code for AST nodes that changed. Or we can remove the original node reference
+        // to the same effect. The code generator will only put parentheses where strictly
+        // necessary, which other rules should handle.
```

Comment:
This can be removed completely, as it's already mentioned in changelog annotation. 
That's the only place that is displayed to output.

Reply:
The changelog only points to the RFC. This comment explains how the rule actually works, which is IME useful for the developer that will be changing it in the future. Commit message is another place that this info could belong to but
1. It is less visible (requires `git blame`)
2. It is immutable (cannot be changed if we change how this is implemented).

Reply:
Should be explained in class docblock then. 
Also, this is not reliable method. IMO this whole rule should be handled in tokens, by phpcsfixer.

Reply:
I would say that explaining how the implementation works is an implementation detail and the docs are targeted at Rector developers. In my understanding, docblocks are for describing the interface targeted for the users of the classes.

If you are aware of more reliable method, it would be appreciated.

You are right that this is a style issue so PHPCSFixer would be a good place to fix it. However, it is much harder to do it on token level since it requires knowing what AST node is given token part of to distinguish the contexts where parentheses are optional from cases where they are required.

On the other hand this is also a downgrade issue. If we do not cover it in a Downgrade set, downgrade will produce syntactically invalid code.

Reply:
This should be part of documentation in `/docs`, we're not putting it to the code by convention.

It's actually easier to do it on token level, as we have it under control. If default printing of nodes changes in `php-parser`, and it probably will with ever new PHP major version, we have a BC break on reprint.  Also we report all those node always changed on every single Rector run, even though untouched.

Reply:
@jtojnar you possibly can check node followed by parentheses via token check, something like `FollowedByCurlyBracketAnalyzer`, but for parentheses, eg:

https://github.com/rectorphp/rector-src/blob/e27e4830944bb418e5b48abc206778be11e20527/rules/Php74/Tokenizer/FollowedByCurlyBracketAnalyzer.php#L10-L19